plugins/e621.py

```diff
@@ -0,0 +1,115 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import logging
+import re
+import html
+from urllib.parse import urlsplit
+import traceback
+
+import requests
+import mimeparse
+import praw
+
+
+class E621Plugin:
+    """
+    Mirrors e621 images using either their API or using their CDN links.
+    Created by /u/HeyItsShuga
+
+    """
+
+    def __init__(self, useragent: str, **options):
+        """Initialize the e621 importer.
+
+        :param useragent: The useragent to use for querying e621.
+        :param options: Other options in the configuration. Ignored.
+        """
+        self.log = logging.getLogger('lapis.e621')
+        self.headers = {'User-Agent': useragent}
+        self.regex = re.compile(
+            r'^https?://(((?:www\.)?(?:static1\.)?'
+            r'((e621)|(e926))\.net/(data/.+/(?P<cdn_id>\w+))?'
+            r'(post/show/(?P<id>\d+)/?)?.*))$')
+
+    def import_submission(self, submission: praw.objects.Submission) -> dict:
+        """Import a submission from e621.
+
+        This function will define the following values in its return data:
+        - author: simply "an anonymous user on e621"
+        - source: The url of the submission
+        - importer_display/header
+        - import_urls
+
+        After we define that, we need to get the image. Since e621 has an
+        API, we use that to try to get the image if the image is a non-CDN URL.
+        If it is a CDN URL, we take the image directory and upload *that* to
+        Imgur.
+
+        image_url is the variable of the image to upload.
+
+        :param submission: A reddit submission to parse.
+        """
+        try:
+            url = html.unescape(submission.url)
+            match = self.regex.match(submission.url)
+            if not match:
+                return None
+            r = requests.head(url, headers=self.headers)
+            mime_text = r.headers.get('Content-Type')
+            mime = mimeparse.parse_mime_type(mime_text)
+            if mime[0] == 'image':
+                self.log.debug('Using the CDN')
+                data = {'author': 'a e926 user',
+                        'source': url,
+                        'importer_display':
+                            {'header': 'Mirrored e926 image:\n\n'}}
+                image_url = url
+            else:
+                self.log.debug('Using the e621 API')
+                match_data = match.groupdict()
+                # For non-CDN links, the plugin attempts to get the ID
+                # out of the URL using regex.
+                id = match_data.get('id')
```

Comment:
`match_data` is unneeded (I realized this recently). `id = match.group('id')` can simply be done. `id` could probably also be named `post_id` to avoid hiding a builtin function.